lib/puppet/functions/lookup.rb

```diff
@@ -0,0 +1,199 @@
+#       `Hash[<value_type>]`, `Hash[<key_type>, <value_type>]`, where `<key_type>`, and
+#       `<value_type>` are the types of the keys and values respectively
+#       (key is `Literal` by default).
+#     * `Data` - abstract type representing any `Literal`, `Array[Data]`, or `Hash[Literal, Data]`
+#     * `Pattern[<p1>, <p2>, ..., <pn>]` - an enumeration of valid patterns (one or more) where
+#        a pattern is a regular expression string or regular expression,
+#        e.g. `Pattern['.com$', '.net$']`, `Pattern[/[a-z]+[0-9]+/]`
+#     * `Enum[<s1>, <s2>, ..., <sn>]`, - an enumeration of exact string values (one or more)
+#        e.g. `Enum[blue, red, green]`.
+#     * `Variant[<t1>, <t2>,...<tn>]` - matches one of the listed types (at least one must be given)
+#       e.g. `Variant[Integer[8000,8999], Integer[20000, 99999]]` to accept a value in either range
+#     * `Regexp`- a regular expression (i.e. the result is a regular expression, not a string
+#        matching a regular expression).
+#   * A string containing a type description - one of the types as shown above but in string form.
+#
+# If the function is called without specifying a default value, and nothing is bound to the given name
+# an error is raised unless the option `accept_undef` is true. If a block is given it must produce an acceptable
+# value (or call `error`). If the block does not produce an acceptable value an error is
+# raised.
+#
+# Examples:
+#
+# When called with one argument; **the name**, it
+# returns the bound value with the given name after having  asserted it has the default datatype `Data`:
+#
+#     lookup('the_name')
+#
+# When called with two arguments; **the name**, and **the expected type**, it
+# returns the bound value with the given name after having asserted it has the given data
+# type ('String' in the example):
+#
+#     lookup('the_name', 'String')
+#     lookup('the_name', String)
+#
+# When called with three arguments, **the name**, the **expected type**, and a **default**, it
+# returns the bound value with the given name, or the default after having asserted the value
+# has the given data type (`String` in the example above):
+#
+#     lookup('the_name', 'String', 'Fred')
+#     lookup('the_name', String, 'Fred')
+#
+# Using a lambda to provide a default value by calling a function:
+#
+#     lookup('the_size', Integer[1,100]) |$name| {
+#       obtain_size_default()
+#     }
+#
+# When using a block, the value it produces is also asserted against the given type, and it may not be
+# `undef` unless the option `'accept_undef'` is `true`.
+#
+# If you want to make lookup return undef when no value was found instead of raising an error:
+#
+#      $are_you_there = lookup('peekaboo', { accept_undef => true} )
+#
+# - Since 4.0.0
+Puppet::Functions.create_function(:lookup, Puppet::Functions::InternalFunction) do
+  name_t = 'Variant[String,Array[String]]'
+  value_type_t = 'Optional[Variant[String,Type]]'
```

Comment:
remove support for type in string format (if we want that in the language it is better to support that with a general mechanism, say `Type["Integer[1,10]"] or a function, (not something we add or do now).